---
title: "Scoring Personalized Molecular Portraits"
author:
- name: Jordi Martorell-Marugán
  affiliation:
  - GENYO, Centre for Genomics and Oncological Research
  email: jordi.martorell@genyo.es
- name: Daniel Toro-Domínguez
  affiliation:
  - GENYO, Centre for Genomics and Oncological Research
  email: daniel.toro@genyo.es
- name: Pedro Carmona-Sáez
  affiliation:
  - Department of Statistics and Operational Research. University of Granada
  - GENYO, Centre for Genomics and Oncological Research
  email: pcarmona@ugr.es
package: pathMED
date: "`r BiocStyle::doc_date()`"
abstract: >
  pathMED uses gene expression data from reference studies to predict phenotypes
  in new datasets (e.g., disease status or response to treatments) at the 
  individual sample level. The 
  workflow is to calculate M-scores (i.e., pathways activities) from both 
  reference and new data. Then, several machine learning methods are evaluated 
  in order to select the best method and to tune hyperparameters. Performance 
  metrics and a ready-to-use model for other datasets are returned.
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Scoring Personalized Molecular Portraits}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
references:
- id: toroDominguez2022
  title: "Scoring personalized molecular portraits identify Systemic Lupus 
  Erythematosus subtypes and predict individualized drug responses, 
  symptomatology and disease progression"
  author:
  - family: Toro-Domínguez
    given: D
  - family: Martorell-Marugán
    given: J
  - family: Martínez-Bueno
    given: M
  - family: López-Domínguez
    given: R
  - family: Carnero-Montoro
    given: E
  - family: Barturen
    given: G
  - family: Goldman
    given: D
  - family: Petri
    given: M
  - family: Carmona-Sáez
    given: P
  - family: Alarcón-Riquelme
    given: ME
  container-title: Briefings in Bioinformatics
  issued:
    year: 2022
  volume: 23
  issue: 5
---

# Introduction
This package contains the functions used in a previous work that used  
personalized scoring to predict the drug response, symptoms and progression of 
Systemic Lupus Erythematosus patients from gene expression data 
@toroDominguez2022.

# Step 1: Prepare the reference data

## 1.1. Calculate the M-scores for reference data

The firs step is to calculate M-scores for the reference datasets. M-scores 
measure the level of disturbance of molecular pathways between patients and 
healthy controls. This approach permits to compare the obtained results with 
the M-scores from new data, even when the measured genes do not completely
overlap.

```{r, message = FALSE, results='hide', warning=FALSE}
library(pathMED)
data(refData, exampleData, exampleMetadata)
```

We loaded to our R environment **refData**, 
**exampleData** and **exampleMetadata** objects. **refData** 
is a list with 4 datasets extracted from NCBI GEO containing lupus and healthy 
samples. For each dataset, there is a gene expression matrix for the patients 
("Disease") and another one for the healthy controls ("Healthy"). You should 
prepare an object with the same structure with as many datasets as possible.
**exampleData** and **exampleMetadata** are the test data for this vignette and 
they will be used in later functions.

```{r, warning=FALSE}
ls(refData)
ls(refData$dataset1)
head(refData$dataset1$Disease)
```

_getMscoresRef()_ function uses this object as input and calculates the 
M-scores for each gene set of the specified database. By default, Reactome
is the used database, but the **genesets** parameter can be modified to use 
KEGG ("kegg"), transcriptional modules ("tmod") or Gene Ontology ("gobp", 
"gomf", "gocc"). Furthermore, is "custom" is selected, any geneset with the 
proper format can be introduced to the **customGeneset** parameter. It should 
be a named list with the genes of each gene set. If you are running the script 
with a multi-core processor, use the **cores** parameter to specify the number 
of workers to be used (most of the pathMED functions include this parameter).

```{r, eval=TRUE, warning=FALSE, message=FALSE, results='hide'}
refMscore <- getMscoresRef(data=refData,
                           genesets="tmod",
                           cores=1)
```

## 1.2. Filter gene sets

It is necessary to find the gene sets that are more relevant for the studied 
phenotype. With the **diseasePaths** function, you can perform such selection 
specifying the percentage of samples that in which the pathways should be 
statistically significant (**perc_samples** parameter) and the minimum number
of datasets that must pass this threshold (**min_datasets** parameter).

```{r, warning=FALSE}
relevantPaths <- diseasePaths(MRef=refMscore,
                       min_datasets=3,
                       perc_samples=10)
```

# Step 2: Calculate the M-scores

Once the reference data is ready, you can calculate the M-scores for your test 
samples, which is a necessary step to construct the machine learning predictors.
**getMscores()** function takes as input an expression matrix with the cases 
samples, an optional expression matrix with healthy controls and the output of 
the **diseasePaths** function. If no healthy controls are provided, the M-scores
are calculated using the *k* most similar samples of the reference data. *k* 
may be specified with the **nk** parameter. The output is a matrix with the 
M-score of each gene set (rows) and sample (columns).

```{r, warning=FALSE}
MScoresExample <- getMscores(genesets = relevantPaths,
                            Patient = exampleData,
                            Healthy = NULL,
                            nk = 5)
MScoresExample[1:5, 1:5]
```

# Step 3: Make predictors

**getML()** function fits and tests several machine learning algorithms in 
order to find which one has the best performance predicting the specified 
variable (**var2predict** parameter). The input objects are the M-scores matrix 
calculated in the step 2 (**expData**), and a data frame with the metadata of 
the samples (**metadata**). One or several algorithms may be selected with the 
**algorithms** parameter (see the function help).

```{r, warning=FALSE, eval=FALSE}
fit.model <- getML(expData=MScoresExample,
                 metadata=exampleMetadata,
                 var2predict="Response",
                 algorithms="svmLinear",
                 outerfolds=10,
                 splitProp=0.8,
                 innerfolds=10,
                 innerRepeats=10,
                 featureFilter = "none",
                 prior="BAUC")
```

# Session info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References
