---
title: "A comprehensive framework for precision medicine with omics data"
author:
- name: Jordi Martorell-Marugán
  affiliation:
  - GENYO, Centre for Genomics and Oncological Research
  - Andalusian Foundation for Biomedical Research in Eastern Andalusia (FIBAO)
  email: jordi.martorell@genyo.es
- name: Daniel Toro-Domínguez
  affiliation:
  - GENYO, Centre for Genomics and Oncological Research
  - Karolinska Institutet
  email: danieltorodominguez@gmail.com
- name: Iván Ellson
  affiliation:
  - GENYO, Centre for Genomics and Oncological Research
- name: Pedro Carmona-Sáez
  affiliation:
  - Department of Statistics and Operational Research. University of Granada
  - GENYO, Centre for Genomics and Oncological Research
  email: pcarmona@ugr.es
package: pathMED
date: "`r BiocStyle::doc_date()`"
abstract: >
  PathMED is a collection of tools to facilitate precision medicine studies
  with omics data (e.g. transcriptomics). Among its funcionalities, genesets
  scores for individual samples may be calculated with several methods.
  These scores may be used to train machine learning models and to predict 
  clinical features on new data. For this, several machine learning
  methods are evaluated in order to select the best method based on internal 
  validation and to tune the hyperparameters. Performance metrics and a 
  ready-to-use model to predict the outcomes for new patients are returned.
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{A comprehensive framework for precision medicine with 
  omics data}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
references:
- id: toroDominguez2022
  title: "Scoring personalized molecular portraits identify Systemic Lupus Erythematosus subtypes and predict individualized drug responses, symptomatology and disease progression"
  author:
  - family: Toro-Domínguez
    given: D
  - family: Martorell-Marugán
    given: J
  - family: Martínez-Bueno
    given: M
  - family: López-Domínguez
    given: R
  - family: Carnero-Montoro
    given: E
  - family: Barturen
    given: G
  - family: Goldman
    given: D
  - family: Petri
    given: M
  - family: Carmona-Sáez
    given: P
  - family: Alarcón-Riquelme
    given: ME
  container-title: Briefings in Bioinformatics
  issued:
    year: 2022
  volume: 23
  issue: 5
- id: hanzelman2013
  title: "GSVA: gene set variation analysis for microarray and RNA-Seq data"
  author:
  - family: Hänzelmann
    given: S
  - family: Castelo
    given: R
  - family: Guinney
    given: J
  container-title: BMC Bioinformatics
  issued:
    year: 2013
  volume: 14
  issue: 7
- id: barbie2009
  title: "Systematic RNA interference reveals that oncogenic KRAS-driven cancers require TBK1"
  author:
  - family: Barbie
    given: DA
  - family: Tamayo
    given: P
  - family: Boehm
    given: JS
  - family: et al
  container-title: Nature
  issued:
    year: 2009
  volume: 462
- id: foroutan2018
  title: "Single sample scoring of molecular phenotypes"
  author:
  - family: Foroutan
    given: H
  - family: Bhuva
    given: DD
  - family: Lyu
    given: R
  - family: et al
  container-title: BMC Bioinformatics
  issued:
    year: 2018
  volume: 19
- id: tomfohr2005
  title: "Pathway level analysis of gene expression using singular value decomposition"
  author:
  - family: Tomfohr
    given: J
  - family: Lu
    given: J
  - family: Kepler
    given: TB
  container-title: BMC Bioinformatics
  issued:
    year: 2005
  volume: 6
- id: lee2018
  title: "Inferring Pathway Activity toward Precise Disease Classification"
  author:
  - family: Lee
    given: E
  - family: Chuang
    given: HY
  - family: Kim
    given: JW
  - family: et al
  container-title: PLOS Computational Biology
  issued:
    year: 2018
  volume: 4
  issue: 11
- id: aibar2017
  title: "SCENIC: single-cell regulatory network inference and clustering"
  author:
  - family: Aibar
    given: S
  - family: González-Blas
    given: CB
  - family: Moerman
    given: T
  - family: et al
  container-title: Nature Methods
  issued:
    year: 2017
  volume: 14
- id: badia2022
  title: "decoupleR: ensemble of computational methods to infer biological activities from omics data"
  author:
  - family: Badia-i-Mompel
    given: P
  - family: Santiago
    given: JV
  - family: Braunger
    given: J
  - family: et al
  container-title: Bioinformatics Advances
  issued:
    year: 2022
  volume: 2
  issue: 1
- id: korotkevich2021
  title: "Fast gene set enrichment analysis"
  author:
  - family: Korotkevich
    given: G
  - family: Sukhov
    given: V
  - family: Budin
    given: N
  - family: et al
  container-title: bioRxiv
editor_options: 
  markdown: 
    wrap: 72
---

# Install pathMED

You can install pathMED from the GitHub repository:

```{r}
devtools::install_github("jordimartorell/pathMED")
```

Please note that on some operating systems such as Ubuntu there may be
installation errors in the packages "metrica", "factoextra" and
"FactoMineR", as it is necessary to install system dependencies with
administrator permissions. These system dependencies can be identified
the following code (adapted to Ubuntu 20.04) and installed by running
the output in a terminal as administrator:

```{r}
pak::pkg_system_requirements(c("metrica", "factoextra", "FactoMineR"), 
                             "ubuntu", "20.04")
```

# 1. Core pipeline

(Add figure)

## 1.1 (Optional) Decomposing gene sets into coexpressed subsets

Gene sets often contain genes associated with a specific biological
function, but their expression can be heterogeneous, sometimes even
exhibiting opposing directions when the function is altered (e.g., in
disease). The *dissectDB* function clusters the genes within each set
into subsets based on coordinated expression patterns. This clustering
can be performed using one or more molecular datasets, which may include
control samples or not. The same dataset used in the following sections
can also be used for this purpose. Both preloaded and custom gene sets
can be used (see the next section for details). The following code
splits KEGG pathways into subpathways using the example dataset:

The output is a gene set that can be used with the getScores function.
In the new gene set, subpathways are labeled with the suffix \_splitx.
For example, for the pathway hsa04714, the resulting subpathways will
include this suffix.

```{r, message = FALSE, results='hide', warning=FALSE}
library(pathMED)
data(exampleData)

customKEGG <- dissectDB(list(exampleData), geneSets = "kegg")
```

The output is a gene set that can be used with the getScores function.
In the new gene set, subpathways are labeled with the suffix splitx. For
example, for the pathway hsa04714:

```{r, message = FALSE, warning=FALSE}
# Before splitting
print(genesetsData[["kegg"]][["hsa04714"]])

# After splitting
print(customKEGG[grep("hsa04714", names(customKEGG))])

```

## 1.2 Scoring

The `getScores()` function calculates a score for each gene set and
sample. It takes as input a matrix or data frame with molecular features
in rows and samples in columns, along with a gene set collection. The
function supports 22 different scoring methods, which can be specified
using the `method` parameter. The available methods and their references
are:

-   M-score (@toroDominguez2022)
-   GSVA (@hanzelman2013)
-   ssGSEA (@barbie2009)
-   singscore (@foroutan2018)
-   Plage (@tomfohr2005)
-   Z-score (@lee2018)
-   AUCell (@aibar2017)
-   ssGSEA (@barbie2009)
-   AUCell (@aibar2017)
-   decoupleR methods (MDT, MLM, ORA, UDT, ULM, WMEAN, norm_WMEAN,
    corr_WMEAN, WSUM, norm_WSUM and corr_WSUM) (@badia2022)
-   FGSEA and norm_FGSEA (@korotkevich2021)

Gene sets can be provided as a named list, where each entry contains the
genes associated with a specific gene set, using the `geneSets`
parameter. Alternatively, preloaded databases can be used by specifying
their name in the `geneSets` parameter:

-   Reactome ("reactome")
-   KEGG ("kegg")
-   Transcriptional modules associated with immune functions ("tmod")
-   Gene Ontology BP, MF and CC ("go_bp", "go_mf", "go_cc")
-   Disgenet ("disgenet")
-   Human Phenotype Ontology ("hpo")
-   WikiPathways ("wikipathways")
-   PharmGKB ("pharmgkb")
-   LINCS ("lincs")
-   Comparative Toxicogenomics Database ("ctd")

If running the script on a multi-core processor, you can use the `cores`
parameter to specify the number of workers (most `pathMED` functions
include this parameter). Additionally, specific parameters for the
scoring functions can be customized (see, for example,
`?GSVA::gsvaParam`).

To run this function with the example data:

```{r, message = FALSE, results='hide', warning=FALSE}
library(pathMED)
data(exampleData)
scoresExample <- getScores(exampleData, geneSets="kegg", method="GSVA")
head(scoresExample)
```

The result is a matrix with gene set identifiers as rows and samples as
columns. To annotate the identifiers with their corresponding
descriptions, you can use the `ann2term()` function.

```{r, message = FALSE, results='hide', warning=FALSE}
annotatedPathways <- ann2term(scoresExample)
head(annotatedPathways)
```

## 1.3. Training ML models

The next step in the core `pathMED` pipeline is to use the scores
calculated in the previous step to train machine learning (ML) models
for predicting categorical or numerical clinical variables. The
`trainModel()` function fits and tests multiple algorithms, selecting
the one with the best performance for predicting the specified variable
(`var2predict` parameter). The input objects are `inputData` (the scores
matrix obtained in the previous step), and `metadata` (a data frame
containing sample metadata).

One or more algorithms can be selected using the `algorithms` parameter
(see the function help for details). Nested cross-validation is
performed using `Koutter` folds for the outer cross-validation (CV) and
`Kinner` folds for the inner CV. The inner k-fold CV can be repeated
multiple times (`repeatsCV` parameter) for hyperparameter tuning.

In the example dataset, the metadata include a column called `Response`,
indicating whether patients responded to a treatment. The
`positiveClass` parameter can be specified to calculate performance
metrics based on the chosen positive class (e.g., diseased).

```{r, warning=FALSE, eval=FALSE}
data(exampleMetadata)

modelsList <- methodsML(algorithms=c('rf', 'knn'), outcomeClass="character")

trainedModel  <- trainModel(scoresExample,
                            metadata=exampleMetadata,
                            var2predict="Response",
                            positiveClass = "YES",
                            models = modelsList,
                            Koutter = 3,
                            Kinner = 3,
                            repeatsCV = 1)
```

# 2. Additional functionalities

# 3. M-Scores

This methodology was used in a previous work that used personalized
scoring to predict the drug response, symptoms and progression of
Systemic Lupus Erythematosus patients from gene expression data
@toroDominguez2022. pathMED has been developed to be applicable to any
disease or biological context.

## Step 1: Prepare the reference data

### 1.1. Build the reference data object

```{r, message = FALSE, results='hide', warning=FALSE}
data(reference_datasets, exampleMetadata)
```

We loaded to our R environment **reference_datasets**, **exampleData**
and **exampleMetadata**. **reference_datasets** includes 4 gene
expression datasets and 4 metadata datasets containing clinical
information for each dataset. These datasets were extracted from NCBI
GEO and contains lupus and healthy samples, codified in the metadata as
"Healthy_sample" and "Disease_sample". **exampleData** and
**exampleMetadata** are the test data for this vignette and they will be
used in later functions.

*buildRefObject()* function can be used to create a refData object with
the structure required for the pathMED pipeline. We can input our
expression datasets as a list (**data**) or as a single dataset if we
only have one. We can introduce a list of **metadata**, one for each
expression dataset or a single metadata with information for samples of
all datasets. The grouping column of metadata (**groupVar**) and the
level of the control group (**controlGroup**), usually healthy samples,
must be introduced (or one for each metadata if they are different).

```{r, message = FALSE, results='hide', warning=FALSE}
refData <- buildRefObject(
  data = list(dataset1, dataset2, dataset3, dataset4),
  metadata = list(metadata1, metadata2, metadata3, metadata4),
  groupVar = "group",
  controlGroup = "Healthy_sample")
```

You should use as many datasets as possible to create the **refData**.
Note that all the gene expression datasets should be log2 transformed
and with a normal distribution. All gene names must be annotated as Gene
Symbol in all reference and test datasets.

### 1.2. Calculate the M-scores for reference data

After creating a **refData** object, the next step is to calculate
M-scores for the reference datasets. M-scores measure the level of
disturbance of molecular pathways between patients and healthy controls.
This approach permits to compare the obtained results with the M-scores
from new data, even when the measured genes do not completely overlap.

*createReference()* function uses this **refData** object as input and
calculates the M-scores for each gene set of the specified database.

```{r, eval=TRUE, warning=FALSE, message=FALSE, results='hide'}
refMscore <- createReference(data=refData,
                           geneSetsDB="tmod",
                           cores=1)
```

### 1.3. Filter gene sets

It is necessary to find the gene sets that are more relevant for the
studied phenotype. With the **diseasePaths** function, you can perform
such selection specifying the percentage of samples that in which the
pathways should be statistically significant (**perc_samples**
parameter) and the minimum number of datasets that must pass this
threshold (**min_datasets** parameter). These two parameters are used to
extract pathways that may be significant in small subgroups of patients
(something very common when dealing with patients with heterogeneous
diseases), but at the same time that these pathways appear deregulated
in different studies (to filter possible biases from individual
studies). The significance of each route in each sample of the different
studies is calculated based on its M-score, as in the original article
@toroDominguez2022.

```{r, warning=FALSE}
relevantPaths <- diseasePaths(MRef=refMscore,
                       min_datasets=3,
                       perc_samples=10)
```

## Step 2: Calculate the M-scores

Once the reference data is ready, you can calculate the M-scores for
your test samples, which is a necessary step to construct the machine
learning predictors. **getMscores()** function takes as input an
expression matrix with the cases samples, an optional expression matrix
with healthy controls and the output of the **diseasePaths** function.
If no healthy controls are provided, the M-scores are calculated using
the *k* most similar samples of the reference data. *k* may be specified
with the **nk** parameter. The output is a matrix with the M-score of
each gene set (rows) and sample (columns).

```{r, warning=FALSE}
MScoresExample <- getMscores(geneSets = relevantPaths,
                            inputData = exampleData)
MScoresExample[1:5, 1:5]
```

## Step 3: Make predictors

# Session info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References
