---
title: "Scoring Personalized Molecular Portraits"
author:
- name: Jordi Martorell-Marugán
  affiliation:
  - GENYO, Centre for Genomics and Oncological Research
  email: jordi.martorell@genyo.es
- name: Daniel Toro-Domínguez
  affiliation:
  - GENYO, Centre for Genomics and Oncological Research
  email: daniel.toro@genyo.es
- name: Pedro Carmona-Sáez
  affiliation:
  - Department of Statistics and Operational Research. University of Granada
  - GENYO, Centre for Genomics and Oncological Research
  email: pcarmona@ugr.es
package: pathMED
date: "`r BiocStyle::doc_date()`"
abstract: >
  pathMED quantifies disease-related molecular disregulations on biological
  pathways for individual samples. First, M-scores (pathways activities) are
  calculated from gene-expression data from reference studies (with samples from
  healthy samples and patients with a certain disease) to extract
  disease-relevant biological pathways. Then, M-scores for the selected pathways
  can be used to predict phenotypes and clinical outcomes in new datasets (e.g., 
  disease status or response to treatments). For this, several machine learning
  methods are evaluated in order to select the best method and to tune
  hyperparameters. Performance metrics and a ready-to-use model to predict the
  trained outcomes for new patients are returned.
output:
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Scoring Personalized Molecular Portraits}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
references:
- id: toroDominguez2022
  title: "Scoring personalized molecular portraits identify Systemic Lupus 
  Erythematosus subtypes and predict individualized drug responses, 
  symptomatology and disease progression"
  author:
  - family: Toro-Domínguez
    given: D
  - family: Martorell-Marugán
    given: J
  - family: Martínez-Bueno
    given: M
  - family: López-Domínguez
    given: R
  - family: Carnero-Montoro
    given: E
  - family: Barturen
    given: G
  - family: Goldman
    given: D
  - family: Petri
    given: M
  - family: Carmona-Sáez
    given: P
  - family: Alarcón-Riquelme
    given: ME
  container-title: Briefings in Bioinformatics
  issued:
    year: 2022
  volume: 23
  issue: 5
---

# Introduction


This methodology was used in a previous work that used 
personalized scoring to predict the drug response, symptoms and progression of 
Systemic Lupus Erythematosus patients from gene expression data 
@toroDominguez2022. pathMED has been developed to be applicable to any disease 
or biological context.

# Step 1: Prepare the reference data

## 1.1. Build the reference data object

```{r, message = FALSE, results='hide', warning=FALSE}
library(pathMED)
data(reference_datasets, exampleData, exampleMetadata)
```

We loaded to our R environment **reference_datasets**, 
**exampleData** and **exampleMetadata**.
**reference_datasets** includes 4 gene expression datasets and 4 metadata 
datasets containing clinical information for each dataset. These datasets 
were extracted from NCBI GEO and contains lupus and healthy samples, codified 
in the metadata as "Healthy_sample" and "Disease_sample".
**exampleData** and **exampleMetadata** are the test data for this vignette and 
they will be used in later functions.

_buildRefData()_ function can be used to create a refData object with the 
structure required for the pathMED pipeline. We can input our expression 
datasets as a list (**data**) or as a single dataset if we only have one. We can 
introduce a list of **metadata**, one for each expression dataset or a single 
metadata with information for samples of all datasets. The grouping column of 
metadata (**groupVar**) and the level of the control group (**controlGroup**), 
usually healthy samples, must be introduced (or one for each metadata if they 
are different).

```{r, message = FALSE, results='hide', warning=FALSE}
refData <- buildRefData(
  data = list(dataset1, dataset2, dataset3, dataset4),
  metadata = list(metadata1, metadata2, metadata3, metadata4),
  groupVar = "group",
  controlGroup = "Healthy_sample")
```

You should use as many datasets as possible to create the **refData**.
Note that all the gene expression datasets should be log2 transformed and with 
a normal distribution. All gene names must be annotated as Gene Symbol in all 
reference and test datasets.

## 1.2. Calculate the M-scores for reference data

After creating a **refData** object, the next step is to calculate M-scores 
for the reference datasets. M-scores measure the level of disturbance of 
molecular pathways between patients and healthy controls. 
This approach permits to compare the obtained results with the M-scores from 
new data, even when the measured genes do not completely overlap.

_getMscoresRef()_ function uses this **refData** object as input and calculates
the M-scores for each gene set of the specified database. By default, Reactome
is the used database, but the **genesets** parameter can be modified to use 
KEGG ("kegg"),  transcriptional modules related with immune functions (“tmod”)
or Gene Ontology ("gobp", 
"gomf", "gocc"). Furthermore, if "custom" is selected, any geneset with the 
proper format can be introduced to the **customGeneset** parameter. It should 
be a named list with the genes of each gene set. If you are running the script 
with a multi-core processor, use the **cores** parameter to specify the number 
of workers to be used (most of the pathMED functions include this parameter).

```{r, eval=TRUE, warning=FALSE, message=FALSE, results='hide'}
refMscore <- getMscoresRef(data=refData,
                           genesets="tmod",
                           cores=1)
```

## 1.3. Filter gene sets

It is necessary to find the gene sets that are more relevant for the studied 
phenotype. With the **diseasePaths** function, you can perform such selection 
specifying the percentage of samples that in which the pathways should be 
statistically significant (**perc_samples** parameter) and the minimum number
of datasets that must pass this threshold (**min_datasets** parameter). These
two parameters are used to extract pathways that may be significant in small 
subgroups of patients (something very common when dealing with patients with 
heterogeneous diseases), but at the same time that these pathways appear 
deregulated in different studies (to filter possible biases from individual 
studies). The significance of each route in each sample of the different studies
is calculated based on its M-score, as in the original article
@toroDominguez2022.

```{r, warning=FALSE}
relevantPaths <- diseasePaths(MRef=refMscore,
                       min_datasets=3,
                       perc_samples=10)
```

# Step 2: Calculate the M-scores

Once the reference data is ready, you can calculate the M-scores for your test 
samples, which is a necessary step to construct the machine learning predictors.
**getMscores()** function takes as input an expression matrix with the cases 
samples, an optional expression matrix with healthy controls and the output of 
the **diseasePaths** function. If no healthy controls are provided, the M-scores
are calculated using the *k* most similar samples of the reference data. *k* 
may be specified with the **nk** parameter. The output is a matrix with the 
M-score of each gene set (rows) and sample (columns).

```{r, warning=FALSE}
MScoresExample <- getMscores(genesets = relevantPaths,
                            Patient = exampleData,
                            Healthy = NULL,
                            nk = 5)
MScoresExample[1:5, 1:5]
```

# Step 3: Make predictors

**getML()** function fits and tests several machine learning algorithms in 
order to find which one has the best performance predicting the specified 
variable (**var2predict** parameter). The input objects are the M-scores matrix 
calculated in the step 2 (**expData**), and a data frame with the metadata of 
the samples (**metadata**). One or several algorithms may be selected with the 
**algorithms** parameter (see the function help). Nested cross-validation is 
performed using k folds for the outter CV (**Koutter**) and inner CV (**Kinner**). 
The inner k-fold cross-validation is repeated several times (**reaptsCV** 
parameter) for hyperparameter tuning.

```{r, warning=FALSE, eval=FALSE}
fit.model <- getML(expData=MScoresExample,
                 metadata=exampleMetadata,
                 var2predict="Response",
                  models = methodsML(algorithms = "all", outcomeClass = "character"),
                  Koutter = 2,
                  Kinner = 3,
                  repeatsCV = 1,
                  continue_on_fail = TRUE,
                  positiveClass = NULL,
                  saveLogFile = NULL)
```

# Session info
```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References
